### PR TITLE
Add @ExtensionMethod support

### LINKED ISSUE #21: Add @ExtensionMethod support
Please add `@ExtensionMethod` support, it is soooo usefull together with libraries like commons-lang3 and guava.

[Bounty](https://www.bountysource.com/issues/3817159-add-extensionmethod-support?utm_source=3817159&utm_medium=shield&utm_campaign=ISSUE_BADGE)

### PR DESCRIPTION
**Pull request by NekoCaffeine**
Fix #21

### CODE REVIEW

**Comment by caoli5288** [comment]
Hi. Seems usage find is bugy.

图片

**Comment by NekoCaffeine** [comment]
> Hi. Seems usage find is bugy.

Indeed, I'll see how to fix it.

**Comment by caoli5288** [comment]
And Generic Type Array is bugy.

图片

**Comment by NekoCaffeine** [comment]
> And Generic Type Array is bugy.

It's a little complicated. I need some time.

File: src/main/java/de/plushnikov/intellij/plugin/hack/extension/EMSupport.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+            final ClassReader reader = new ClassReader(bytes);
+            reader.accept(node, 0);
+            for (final MethodNode method : node.methods) {
+              if (method.name.equals("processDeclarationsInClass") && method.desc.equals("(Lcom/intellij/psi/PsiClass;Lcom/intellij/psi/scope/PsiScopeProcessor;Lcom/intellij/psi/ResolveState;Ljava/util/Set;Lcom/intellij/psi/PsiElement;Lcom/intellij/psi/PsiElement;Lcom/intellij/pom/java/LanguageLevel;ZLcom/intellij/psi/search/GlobalSearchScope;)Z")) {
+                final InsnList instructions = method.instructions;
+                int count = 2;
+                final List<AbstractInsnNode> target = new ArrayList<>();
+                for (final ListIterator<AbstractInsnNode> iterator = instructions.iterator(instructions.size()); iterator.hasPrevious() && count > 0; ) {
+                  final AbstractInsnNode insn = iterator.previous();
+                  if (insn.getOpcode() == IRETURN) {
```
Comment (mplushnikov) [inline src/main/java/de/plushnikov/intellij/plugin/hack/extension/EMSupport.java:65]: Ok, as I see, you inject yourself before every return statement. What do you think would be a best place to add normal ExtensionPoint to this method? If you want, provide a patch for this and I'l look how to apply it to intelliJ sources to remove need for bytecode instrumentation.

**Comment by mplushnikov** [comment]
Good work @NekoCaffeine !

**Comment by NekoCaffeine** [comment]
> And Generic Type Array is bugy.

@caoli5288 
image
That looks better, doesn't it?

And I have considered other special circumstances.
Just like that.
image
Very safe type inference that does not allow runtime class cast exceptions like static inference does.

File: src/main/java/de/plushnikov/intellij/plugin/hack/extension/EMSupport.java (hunk 1 above)
Comment (NekoCaffeine) [reply src/main/java/de/plushnikov/intellij/plugin/hack/extension/EMSupport.java:65]: Putting the extension point in PsiClassImplUtil#processDeclarationsInClass is the simplest solution, just pass the arguments of this method to the corresponding method of the extension point as well.

**Comment by mplushnikov** [comment]
@NekoCaffeine Please add yourself to the list of contributors in Readme.md file

**Comment by NekoCaffeine** [comment]
@mplushnikov Okey-dokey

File: src/main/java/de/plushnikov/intellij/plugin/hack/extension/EMHandler.java [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+      .filter(e -> checkType(e.getKey(), type, node))
+      .map(Map.Entry::getValue)
+      .flatMap(List::stream)
+      .map(function -> {
+        try {
+          return function.apply(node, type);
+        } catch (Throwable throwable) {
+          if (throwable instanceof ProcessCanceledException)
+            throw throwable;
+          throwable.printStackTrace();
```
Comment (mplushnikov) [inline src/main/java/de/plushnikov/intellij/plugin/hack/extension/EMHandler.java:125]: I think to use "printStackTrace" is not a very optimal  solution

**Comment by mplushnikov** [comment]
@NekoCaffeine Just tried it out, looks working pretty well! 

But what do you think, is it possible to make this configurable by some setting option? I want to make it disabled by default and only activate if a user set corresponding flag in the ui to test it less invasive.

Paralllel l ask Jetbrains about adding corresponding ExtensionPoint to `com.intellij.psi.impl.PsiClassImplUtil#processDeclarationsInClass(com.intellij.psi.PsiClass, com.intellij.psi.scope.PsiScopeProcessor, com.intellij.psi.ResolveState, java.util.Set<com.intellij.psi.PsiClass>, com.intellij.psi.PsiElement, com.intellij.psi.PsiElement, com.intellij.pom.java.LanguageLevel, boolean, com.intellij.psi.search.GlobalSearchScope)` method.

File: src/main/java/de/plushnikov/intellij/plugin/hack/extension/EMHandler.java (hunk 2 above)
Comment (NekoCaffeine) [reply src/main/java/de/plushnikov/intellij/plugin/hack/extension/EMHandler.java:125]: Oh, yes, this try block I should remove, actually no exceptions occur here, it's a legacy I used to test cache consistency before.

**Comment by NekoCaffeine** [comment]
> But what do you think, is it possible to make this configurable by some setting option? I want to make it disabled by default and only activate if a user set corresponding flag in the ui to test it less invasive.

Failure of the injection attempt does not result in any consequences, only the invalidation of the function.
So I think a switch is not very useful.

> Can you also add some unit tests to automatically test your solution? At the moment you test it manually I think.

Yes, I'll write the automated test cases later.

**Comment by mplushnikov** [comment]
> Failure of the injection attempt does not result in any consequences, only the invalidation of the function.
> So I think a switch is not very useful.
I think more about possible performance or general compatibility issues, do a quick kill switch would be useful.

**Comment by NekoCaffeine** [comment]
> I think more about possible performance or general compatibility issues, do a quick kill switch would be useful.

In that case the switch should be on by default? (i.e., the feature is supported by default)

**Comment by NekoCaffeine** [comment]
image
Like this?

**Comment by caoli5288** [comment]
Hi. bugy code analyze here.

图片

图片

**Comment by NekoCaffeine** [comment]
@caoli5288 
image

**Comment by NekoCaffeine** [comment]
gif
 
Unknowingly it was four o'clock in the morning.

**Comment by NekoCaffeine** [comment]
image
I don't think there should be any problems with the generic type at this point.

**Comment by NekoCaffeine** [comment]
@mplushnikov Are you waiting for an official API?

**Comment by mplushnikov** [comment]
@NekoCaffeine, I'm sorry about delay, there is a lot to do for me at the moment. 

But I have finally spoken with Jetbrains cowokers, specially with Anna Kozlova. She is not very happy about hacking this inside of `processDeclarationsInClass` of `PsiClassImplUtil`.  There are some risks to break something excpecially for big enterprise project. 

But we can try to do it for the next IntelliJ version 2021, so it should be enougth time to get it well tested!

There some steps to be done:
1.  Please provide new separate PR for Intellij **community** to add new ExtensionPoint and this feature inside of bundled lombok plugin. Follow process described here: https://confluence.jetbrains.com/display/IDEA/GitHub+Pull+Requests

2.  I can merge this PR into my repository and make a new build with your implementation, but I prefer to make it **disabled by default**, to reduce risk to other plugin user. This would be availabel in IntelliJ 2020.2 and maybe some versions before. We can look together, to port it to some of the previos versions to. At the moment I provide support for IntelliJ 2018.1-2020.2 version.

3. **The hardest part**: Coming IntelliJ version 2020.3 is EAP at the moment. It's already feature frezed and we can not add this change to it. This is also the first IntellliJ version with my lombok plugin bundeled by default. So here we can only try to provide custom build of the plugin (or a separate one, because your changes are very easy to separate from the base), and a user would need to install it manually instead of (or together with)  bundeled and provided by default.

What do you think about this solutions?

**Comment by NekoCaffeine** [comment]
Okay, let's take a look at the `build number`.
https://jetbrains.org/intellij/sdk/docs/basics/getting_started/build_number_ranges.html

`181` -> `2018.1`
`203` -> `2020.3`

`181`: https://github.com/JetBrains/intellij-community/blob/181/java/java-psi-impl/src/com/intellij/psi/impl/PsiClassImplUtil.java#L456
`203`: https://github.com/JetBrains/intellij-community/blob/203/java/java-psi-impl/src/com/intellij/psi/impl/PsiClassImplUtil.java#L402

Let's do a simple `diff`.
image
The `only difference` is a final keyword.
I said I chose a method signature that `hasn't changed in a long time` as a entry point.
This injection, even if it `failed`, would have `no consequences` at all.

If you are not satisfied with this, you can change the behavior to `not on` by default.
Did we inject it? Yes.
Under what circumstances would we have the method handle more elements? Only if @ExtensionMethod exists.

And you say you want to support `2018.1`, but I downloaded `2018.1` and found that you wrote `since-build="201.00"`
Maybe your `IDEA_VERSION` is `181`.
But your build script wrote `plugins 'java'`.
That's funny, if your test environment is `2018.1` then you will encounter an exception with gradle sync.
https://blog.jetbrains.com/platform/2019/06/java-functionality-extracted-as-a-plugin/

```stacktrace
> Task :generateLombokConfigParser FAILED
Exception in thread "main" java.lang.NoSuchMethodError: com.intellij.mock.MockApplication.getExtensionArea()Lcom/intellij/openapi/extensions/impl/ExtensionsAreaImpl;
	at org.intellij.grammar.LightPsi$Init.initExtensions(LightPsi.java:211)
	at org.intellij.grammar.LightPsi$MyParsing.<init>(LightPsi.java:169)
	at org.intellij.grammar.LightPsi.<clinit>(LightPsi.java:66)
	at org.intellij.grammar.Main.main(Main.java:49)
```
ha ha!
Okay, I'll try manually modifying `plugin.xml` after the build.

```groovy
sourceCompatibility = JavaVersion.VERSION_11
targetCompatibility = JavaVersion.VERSION_11
```
Hey, why would you say you still support `2018.1`?
```
While loading class de.plushnikov.intellij.plugin.language.LombokConfigCompletionContributor: 
de/plushnikov/intellij/plugin/language/LombokConfigCompletionContributor
 has been compiled by a more recent version of the Java Runtime (class file version 55.0),
 this version of the Java Runtime only recognizes class file versions up to 52.0 [Plugin: Lombook Plugin]
```
The JBR corresponding to `2018.1` is `Java8`.

I used `Java8+` syntax, but of course it was very simple to fix.

```stacktrace
com.intellij.openapi.application.Application.getService(Ljava/lang/Class;)Ljava/lang/Object;
java.lang.NoSuchMethodError: com.intellij.openapi.application.Application.getService(Ljava/lang/Class;)Ljava/lang/Object;
	at de.plushnikov.intellij.plugin.settings.LombokSettings.getInstance(LombokSettings.java:24)
	at de.plushnikov.intellij.plugin.activity.LombokPluginUpdateActivity.runActivity(LombokPluginUpdateActivity.java:19)
	at com.intellij.ide.startup.impl.StartupManagerImpl.runAndMeasure(StartupManagerImpl.java:176)
	at com.intellij.ide.startup.impl.StartupManagerImpl.logActivityDuration(StartupManagerImpl.java:162)
	at com.intellij.ide.startup.impl.StartupManagerImpl.lambda$runPostStartupActivitiesFromExtensions$1(StartupManagerImpl.java:150)
	at com.intellij.ide.startup.impl.StartupManagerImpl.runActivity(StartupManagerImpl.java:368)
	at com.intellij.ide.startup.impl.StartupManagerImpl.runPostStartupActivitiesFromExtensions(StartupManagerImpl.java:152)
	at com.intellij.openapi.project.impl.ProjectManagerImpl.lambda$openProject$8(ProjectManagerImpl.java:397)
	at com.intellij.openapi.project.impl.ProjectManagerImpl.loadProjectUnderProgress(ProjectManagerImpl.java:428)
	at com.intellij.openapi.project.impl.ProjectManagerImpl.openProject(ProjectManagerImpl.java:411)
	at com.intellij.openapi.project.impl.ProjectManagerImpl$2.compute(ProjectManagerImpl.java:488)
	at com.intellij.openapi.project.impl.ProjectManagerImpl$2.compute(ProjectManagerImpl.java:479)
	at com.intellij.openapi.progress.Task$WithResult.run(Task.java:320)
	at com.intellij.openapi.progress.impl.CoreProgressManager$TaskRunnable.run(CoreProgressManager.java:750)
	at com.intellij.openapi.progress.impl.CoreProgressManager$5.run(CoreProgressManager.java:434)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$1(CoreProgressManager.java:157)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:580)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:525)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:85)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:144)
	at com.intellij.openapi.application.impl.ApplicationImpl.lambda$null$10(ApplicationImpl.java:565)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.run(ApplicationImpl.java:305)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```
Are you serious when you say you support `2018.1`?
Are you serious?
The runtime has so many errors that it is impossible to test at all.

**Comment by NekoCaffeine** [comment]
I'm so angry right now, I just want to close this pr and sail away after. FUCK MY LIFE.

**Comment by mplushnikov** [comment]
@NekoCaffeine I think you misunderstood me in point 2 :-( You absolutly don't need to make a 2018.1 working solution right now. You can help, but don't must. And this is something totally irrelevant at this point. 
I have only explained what are the versions I'm dealing with. The current master in this repository is supposed to support 2020.2 version only. All other versions are supported by other "backport" branches, like https://github.com/mplushnikov/lombok-intellij-plugin/tree/backport_intellij_2018.1 for example.

The main step I tell you is to create a separate PR for https://github.com/JetBrains/intellij-community, if you want to get it working for comming IntelliJ versions, because you are the main contributor for this feauture!

**Comment by NekoCaffeine** [comment]
I will when I've calmed down for a while.

**Comment by NekoCaffeine** [comment]
To make a long story short, as shown in the figure, you should be able to backport this solution directly to 2018.1.

**Comment by NekoCaffeine** [comment]
@mplushnikov Are you satisfied with this(28a4d4cad4d5c2f4a18e9f1905724797780db8cf) change?

**Comment by mplushnikov** [comment]
@NekoCaffeine I'm very appreciate how hard you work to get it working well! But did you made PR for you implementation in https://github.com/JetBrains/intellij-community project? This is the main task we need to pass by to get it working with future IntelliJ versions.

**Comment by NekoCaffeine** [comment]
@mplushnikov I'm thinking about how to design the extension points and will submit the pr around tomorrow.

**Comment by NekoCaffeine** [comment]
> (You may be required to reopen your project to take this change correctly applied.)

I think you may want to consider removing this tip.

**Comment by mplushnikov** [comment]
> I think you may want to consider removing this tip.
Yes, I'm working now on completely rework of lombok settings page.

**Comment by NekoCaffeine** [comment]
I'm tested Windows (native Windows 10), Linux (virtual machine Ubuntu 20.10), and MacOS (virtual machine macOS Catalina 10.15). All can be used directly after installing the plugin.

**Comment by NekoCaffeine** [comment]
By the way, you don't have to worry about platform compatibility, there is no native code, just calls to the JVM built-in functions.

The problem on MacOS should be considered a JNA bug at heart.

Also, the injection process is instantaneous and has no subsequent effect.

I've thought about the performance impact and used multi-layered caching and lazy loading, so I don't think you need to think about performance either.

And I don't think this feature has any compatibility or performance conflicts with other plugins.

In summary, I recommend that you enable this feature by default to avoid user confusion.

**Comment by NekoCaffeine** [comment]
@mplushnikov Hey, I've noticed that some annotations(e.g. Delegate) don't update their status after restarting the check engine, and even code changes don't change instantly.
Seems like there's some problem with the cache. I will draft another pr to fix this later.

**Comment by NekoCaffeine** [comment]
@mplushnikov  Hey, I totally fixed the issue with configuration changes taking effect instantly.
GIF 2020-11-8 3-59-20

**Comment by mplushnikov** [comment]
>  Hey, I totally fixed the issue with configuration changes taking effect instantly.

Thank you, but a new PR for this would be better...

**Comment by NekoCaffeine** [comment]
> Thank you, but a new PR for this would be better...

@mplushnikov I was thinking the same thing, but I realized the changes weren't that big.

**Comment by mplushnikov** [comment]
@NekoCaffeine 
I just pushed you changes down to plugin branches for previous versions of IntelliJ. It looks good for Intellij >=2019.2!

For versions 2018.1-2019.1 it doesn't work. If you want, maybe you can look on it. It's here: https://github.com/mplushnikov/lombok-intellij-plugin/tree/backport_intellij_2018.1

**Comment by NekoCaffeine** [comment]
@mplushnikov 
Sorry, I've been so busy lately that I didn't see your reply.
I will find time later to see how to make it compatible with earlier versions.
A belated Happy New Year to you.
